### ISSUE TITLE
Not receiving first view Retained messages after subscribing

### ISSUE THREAD

**Issue by gilbertsmink** [sbtinstruments/asyncio-mqtt#5]
I have implemented the example “Advance use”. However, after subscribing do not receive all retained messages with _log_filtered_messages_.

With both your example and MQTTBox (Chrome browser extention) I subscribe to the same broker and topic ‘#’, and expect to receive 4 retained messages. With MQTTbox I receive all 4 messages, but with your client I only receive 2 retained messages. I tried looking into your code to explore what could be the issue, but lag the experience to debug.

My code for subscribing and handling the message:
```
       await client.subscribe('#')
        asyncio.create_task(log_filtered_messages(client, '#'))
        asyncio.create_task(log_unfiltered_messages(client))
```

What I managed to do is add debug logging and I have created the log below. In yellow highlighted the messaged that are received, but are not handled with my script.

```
2020-05-14 11:03:50,534 - root - DEBUG - Sending CONNECT (u1, p1, wr0, wq0, wf0, c1, k60) client_id=b''
2020-05-14 11:03:50,538 - root - DEBUG - Received CONNACK (0, 0)
2020-05-14 11:03:50,539 - root - DEBUG - Sending SUBSCRIBE (d0, m1) [(b'#', 0)]
2020-05-14 11:03:50,541 - root - DEBUG - Received SUBACK
2020-05-14 11:03:50,542 - root - DEBUG - Received PUBLISH (d0, q0, r1, m0), 'modules/...', ...  (421 bytes)
2020-05-14 11:03:50,544 - root - DEBUG - Received PUBLISH (d0, q0, r1, m0), 'modules/....', ...  (524 bytes)
2020-05-14 11:03:50,545 - root - DEBUG - Received PUBLISH (d0, q0, r1, m0), 'modules/...', ...  (498 bytes)
[topic_filter="#"]: MESSAGE 1
2020-05-14 11:03:50,546 - root - DEBUG - Received PUBLISH (d0, q0, r1, m0), 'modules/...', ...  (469 bytes)
[topic_filter="#"]: MESSAGE 2
```

Have you experienced this behavior before, or have an idea what is going wrong?

**Comment by gilbertsmink** [comment]
I actually found a solution, but lag experience to solve this in your code.

When I use mqtt_c._client.subscribe('#') I receive all 4 retained messages.

**Comment by frederikaalund** [comment]
Thanks for reporting this issue. In order to help you, I need you to first make [a reproducible example](https://stackoverflow.com/help/minimal-reproducible-example).

Your current example is not self-contained. That is, I can't run it from here, since I don't have:
 * The complete source code
 * The broker that you are connecting to

---

All that being said, I did a quick test to try and reproduce the issue anyhow. :) I used the following code:
```
import asyncio
from random import randrange
from asyncio_mqtt import Client

async def log_filtered_messages(client, topic_filter):
    async with client.filtered_messages(topic_filter) as messages:
        async for message in messages:
            topic = message.topic
            value = message.payload.decode()
            print(f'[topic_filter="{topic_filter}", topic={topic}]: {value}')

async def log_unfiltered_messages(client):
    async with client.unfiltered_messages() as messages:
        async for message in messages:
            print(f'[unfiltered]: {message.payload.decode()}')

async def main():
    async with Client('zeus2002003') as client:
        await client.subscribe('#')

        # You can create any number of message filters
        asyncio.create_task(log_filtered_messages(client, '#'))

        # All messages that doesn't match a filter will get logged here
        asyncio.create_task(log_unfiltered_messages(client))

        # Publish a random value to each of these topics
        topics = [
            'something/topic0',
            'something/topic1',
            'something/topic2',
            'something/topic3',
            'something/topic4',
            'something/topic5',
        ]
        while True:
            for topic in topics:
                message = randrange(100)
                print(f'[topic="{topic}"] Publishing message={message}')
                await client.publish(topic, message, qos=1, retain=True)
                await asyncio.sleep(0.1)

asyncio.run(main())
```

I connect to a Mosquitto MQTT broker running on another computer on the local network with hostname `zeus2002003`.

I first make an initial run to set the retained messages. Then, I run the code again and get the following output:
```
[topic="something/topic0"] Publishing message=83
[topic_filter="#", topic=floors/rooftop/illuminance]: 54
[topic_filter="#", topic=something/topic0]: 32
[topic_filter="#", topic=something/topic1]: 54
[topic_filter="#", topic=something/topic2]: 26
[topic_filter="#", topic=something/topic3]: 74
[topic_filter="#", topic=something/topic4]: 73
[topic_filter="#", topic=something/topic5]: 70
[topic_filter="#", topic=something/topic0]: 83
[topic="something/topic1"] Publishing message=20
[topic_filter="#", topic=something/topic1]: 20
[topic="something/topic2"] Publishing message=25
[topic_filter="#", topic=something/topic2]: 25
[topic="something/topic3"] Publishing message=23
[topic_filter="#", topic=something/topic3]: 23
[topic="something/topic4"] Publishing message=14
[topic_filter="#", topic=something/topic4]: 14
[topic="something/topic5"] Publishing message=25
[topic_filter="#", topic=something/topic5]: 25
[topic="something/topic0"] Publishing message=97
[topic_filter="#", topic=something/topic0]: 97
[topic="something/topic1"] Publishing message=84
[topic_filter="#", topic=something/topic1]: 84
[topic="something/topic2"] Publishing message=4
[topic_filter="#", topic=something/topic2]: 4
[topic="something/topic3"] Publishing message=80
[topic_filter="#", topic=something/topic3]: 80
[topic="something/topic4"] Publishing message=20
[topic_filter="#", topic=something/topic4]: 20
...
```
Note how all of the retained messages are instantly received in the beginning, as expected.

In summary, I can't reproduce the issue, sorry. Please provide a reproducible code sample so that we can get to the bottom of this. :)

**Comment by frederikaalund** [comment]
Please also consider the QoS level of the retained message. Messages with a low QoS level can (and will) get lost.

**Comment by gilbertsmink** [comment]
I tried to create a reproducible example with the mosquito test server, but cannot reproduce it there.. I was wondering if it could be a timing issue. The broker (also mosquito) is on my machine in a virtual box, which is much faster. 

```
import asyncio, logging, sys
from asyncio_mqtt import Client as MQTT_Client

async def mqtt_producer(logger):
    logger.info("Created new MQTT producer..")
    
    async with MQTT_Client('localhost', port=1883, logger=logger) as mqtt_c:
        
        logger.info("Connected MQTT client..")
        
        logger.info("Subscribing to #..")
        #mqtt_c._client.subscribe('#')
        await mqtt_c.subscribe('#')
        logger.info("Done subscribing to #..")
        
        while True:
            
            async with mqtt_c.filtered_messages('#') as messages:
                async for message in messages:
                    
                    print('-----------')
                    print(message.payload.decode())
                    print('-----------')
                
            await asyncio.sleep(0)
        
if __name__ == "__main__":
    
    logger = logging.getLogger()
    logger.setLevel(logging.DEBUG)

    handler = logging.StreamHandler(sys.stdout)
    handler.setLevel(logging.DEBUG)
    formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')
    handler.setFormatter(formatter)
    logger.addHandler(handler)
    
    # get event loop
    loop = asyncio.get_event_loop()
    # create producer task
    loop.create_task(mqtt_producer(logger))
```

These logs might help, since they show a message is already received before the log message that the subscription is done.

```
2020-05-14 16:12:13,715 - root - INFO - Created new MQTT producer..
2020-05-14 16:12:14,743 - root - DEBUG - Sending CONNECT (u1, p1, wr0, wq0, wf0, c1, k60) client_id=b''
2020-05-14 16:12:14,748 - root - DEBUG - Received CONNACK (0, 0)
2020-05-14 16:12:14,753 - root - INFO - Connected MQTT client..
2020-05-14 16:12:14,756 - root - INFO - Subscribing to #..
2020-05-14 16:12:14,758 - root - DEBUG - Sending SUBSCRIBE (d0, m1) [(b'#', 0)]
2020-05-14 16:12:14,762 - root - DEBUG - Received SUBACK
2020-05-14 16:12:14,765 - root - DEBUG - Received PUBLISH (d0, q0, r1, m0), 'modules/..', ...  (498 bytes)
2020-05-14 16:12:14,767 - root - INFO - Done subscribing to #..
2020-05-14 16:12:14,768 - root - DEBUG - Received PUBLISH (d0, q0, r1, m0), 'modules/..', ...  (564 bytes)
-----------
PRINTED PAYLOAD HERE..
-----------
2020-05-14 16:12:14,770 - root - DEBUG - Received PUBLISH (d0, q0, r1, m0), 'modules/..', ...  (421 bytes)
-----------
PRINTED  PAYLOAD HERE..
-----------
2020-05-14 16:12:14,772 - root - DEBUG - Received PUBLISH (d0, q0, r1, m0), 'modules/..', ...  (364 bytes)
-----------
PRINTED PAYLOAD HERE..
-----------
2020-05-14 16:13:14,758 - root - DEBUG - Sending PINGREQ
2020-05-14 16:13:14,759 - root - DEBUG - Received PINGRESP
2020-05-14 16:14:14,792 - root - DEBUG - Sending PINGREQ
2020-05-14 16:14:14,793 - root - DEBUG - Received PINGRESP
2020-05-14 16:15:14,825 - root - DEBUG - Sending PINGREQ
2020-05-14 16:15:14,830 - root - DEBUG - Received PINGRESP
```

**Comment by frederikaalund** [comment]
Thanks for providing the full source code.

> I tried to create a reproducible example with the mosquito test server, but cannot reproduce it there.. I was wondering if it could be a timing issue. The broker (also mosquito) is on my machine in a virtual box, which is much faster.

That is a very interesting observation! :)

It could very well be due to network issues. Remember that the [QoS level affects how the message is delivered](https://www.hivemq.com/blog/mqtt-essentials-part-6-mqtt-quality-of-service-levels/). Have you tried specifying `qos=2` on the retained messages? Note that paho's default QoS level is 0, which means that the message will be delivered "at most once", i.e., "message may not be delivered".

**Comment by frederikaalund** [comment]
I was looking through your original e-mail again and thought of something.

Hypothesis: It's due to a [race condition](https://en.wikipedia.org/wiki/Race_condition) between:
 1. when you subscribe to the "#" topic
 2. when you start start listening for messages

Right now, you first do step 1 and then do step 2. Try reversing these steps in your code.

That is, try to do:

```
import asyncio
from random import randrange
from asyncio_mqtt import Client

async def log_filtered_messages(client, topic_filter):
    async with client.filtered_messages(topic_filter) as messages:
        async for message in messages:
            topic = message.topic
            value = message.payload.decode()
            print(f'[topic_filter="{topic_filter}", topic={topic}]: {value}')

async def log_unfiltered_messages(client):
    async with client.unfiltered_messages() as messages:
        async for message in messages:
            print(f'[unfiltered]: {message.payload.decode()}')

async def publish_to_topics(client, topics):
    while True:
        for topic in topics:
            message = randrange(100)
            print(f'[topic="{topic}"] Publishing message={message}')
            await client.publish(topic, message, qos=1, retain=True)
            await asyncio.sleep(0.1)

async def main():
    async with Client('zeus2002003') as client:
        # Step 2: Start listening for messages.
        # You can create any number of message filters
        asyncio.create_task(log_filtered_messages(client, '#'))

        # All messages that doesn't match a filter will get logged here
        asyncio.create_task(log_unfiltered_messages(client))

        # Publish a random value to each of these topics
        topics = [
            'something/topic0',
            'something/topic1',
            'something/topic2',
            'something/topic3',
            'something/topic4',
            'something/topic5',
        ]
        asyncio.create_task(publish_to_topics(client, topics))

        # Wait a while to ensure that we start listening (step 2) before we
        # subscribe to the topic (step 1)
        await asyncio.sleep(1)

        # Step 1: Subscribe to the "#" topic
        await client.subscribe('#')

asyncio.run(main())
```

Let me know if that works for you. :)

**Comment by gilbertsmink** [comment]
That is it, nice find! Thanks a lot for thinking with me

Seems this is also the way it is described in the Paho example: https://github.com/eclipse/paho.mqtt.python/blob/master/examples/client_sub-multiple-callback.py

**Comment by gilbertsmink** [comment]
Will you change your example codes?

**Comment by frederikaalund** [comment]
> That is it, nice find! Thanks a lot for thinking with me

Great! I'm glad that it worked for you. :)

> Seems this is also the way it is described in the Paho example: https://github.com/eclipse/paho.mqtt.python/blob/master/examples/client_sub-multiple-callback.py

Yes, same ordering between the steps.

> Will you change your example codes?

Yes, I will. Even though there is technically nothing wrong with the examples as is, we might as well avoid future confusion.

Thanks again for raising this issue. :+1: Let me know if you run into other issues.

I'll leave this issue open as a reminder until I get the time to adjust the example.

**Comment by frederikaalund** [comment]
I've updated the examples in `README.md`.
